File: Realtime Raytracing/python/03 planes/finished/material.py
from config import *

class Material:
        
    def __init__(self, width: int, height: int):
    
        self.texture = glGenTextures(1)
        glActiveTexture(GL_TEXTURE0)
        glBindTexture(GL_TEXTURE_2D, self.texture)

        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST)

    
        glTexStorage2D(GL_TEXTURE_2D, 1, GL_RGBA32F, width, height)
    
    def writeTo(self) -> None:

        glActiveTexture(GL_TEXTURE0)
        glBindImageTexture(0, self.texture, 0, GL_FALSE, 0, GL_WRITE_ONLY, GL_RGBA32F)

    def readFrom(self) -> None:

        glActiveTexture(GL_TEXTURE0)
        glBindTexture(GL_TEXTURE_2D, self.texture)
    
    def destroy(self) -> None:

        glDeleteTextures(1, (self.texture,))

File: APIs/OpenGL Shader Programming/06 Silhouette Effect/adjacency_triangles/src/view/material.cpp
#include "material.h"

Material::Material(MaterialCreateInfo* createInfo) {
	int texWidth, texHeight;
	image material = util::load_from_file(createInfo->filename);
	texWidth = material.width;
	texHeight = material.height;
	unsigned char* data = material.pixels;
	glCreateTextures(GL_TEXTURE_2D, 1, &texture);
	glTextureStorage2D(texture, 1, GL_RGBA8, texWidth, texHeight);
	glTextureSubImage2D(texture, 0, 0, 0, texWidth, texHeight, GL_RGBA, GL_UNSIGNED_BYTE, data);
	glTextureParameteri(texture, GL_TEXTURE_WRAP_S, GL_REPEAT);
	glTextureParameteri(texture, GL_TEXTURE_WRAP_T, GL_REPEAT);
	glTextureParameteri(texture, GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_LINEAR);
	glTextureParameteri(texture, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
	util::free_image_memory(material);
}

Material::~Material() {
	glDeleteTextures(1, &texture);
}
void Material::use() {
	glBindTextureUnit(0, texture);
}